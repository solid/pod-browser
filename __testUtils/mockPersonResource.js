/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  asUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf, space } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";
import { vcardExtras } from "../src/addressBook";

const VCARD_WEBID_PREDICATE = vcardExtras("WebId");

export function mockWebIdNode(
  webId,
  nodeUrl = "https://example.org/contacts/Person/1234/index.ttl#4567"
) {
  const webIdNode = chain(
    mockThingFrom(nodeUrl),
    (t) => addUrl(t, rdf.type, VCARD_WEBID_PREDICATE),
    (t) => addUrl(t, vcard.value, webId)
  );
  const url = asUrl(webIdNode);
  return { webIdNode, webIdNodeUrl: url };
}

export const aliceProfileUrl = "http://alice.example.com/alice";
export const aliceWebIdUrl = "http://alice.example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://alice.example.com/alice.jpg";
export const alicePodRoot = "http://alice.example.com/";
export const aliceAlternativeProfileUrl = "https://alice2.example.org/card";
export const aliceAlternativeWebIdUrl = "https://alice2.example.org/card#me";

export function mockPersonThingAlice(...operations) {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, vcard.url, aliceAlternativeWebIdUrl),
    (t) => addUrl(t, space.storage, alicePodRoot),
    ...operations
  );
}

export function mockPersonDatasetAlice(...operations) {
  return setThing(
    mockSolidDatasetFrom(aliceWebIdUrl),
    mockPersonThingAlice(...operations)
  );
}

export function mockProfileAlice(...operations) {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice(...operations));
}

export const bobProfileUrl = "http://bob.example.com/bob";
export const bobWebIdUrl = "http://bob.example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";
export const bobPodRoot = "http://bob.example.com/";
export const bobAlternateProfileUrl = "https://bob2.example.org/card";
export const bobAlternateWebIdUrl = "https://bob2.example.org/card#me";

export function mockPersonThingBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, bobAlternateWebIdUrl),
    (t) => addUrl(t, space.storage, bobPodRoot)
  );
}

export function mockPersonDatasetBob() {
  return setThing(mockSolidDatasetFrom(bobWebIdUrl), mockPersonThingBob());
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
