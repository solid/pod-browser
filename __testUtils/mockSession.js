/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";
import mockResponse from "./mockResponse";

import profileTurtle from "./mocks/profile.ttl";
import storageTurtle from "./mocks/podRoot.ttl";
import storageAclTurtle from "./mocks/storageAcl.ttl";

export const webIdUrl = "http://example.com/webId#me";
export const storageUrl = "http://example.com/";
export const storageAclUrl = "http://example.com/.acl";
export const anotherUsersStorageUrl = "http://anotheruser.com/";
export const anotherUsersStorageAclUrl = "http://anotheruser.com/.acl";

export { profileTurtle };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  const { webId, ...restOptions } = options;
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webId || webIdUrl]: () =>
        mockResponse(200, profileTurtle, { "Content-Type": "text/turtle" }),
      [storageUrl]: () =>
        mockResponse(200, storageTurtle, {
          Link: `<${storageAclUrl}>; rel="acl"`,
          "Content-Type": "text/turtle",
        }),
      [storageAclUrl]: () =>
        mockResponse(200, storageAclTurtle, { "Content-Type": "text/turtle" }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: options.webId || webIdUrl,
    }, // add more properties as needed
    ...restOptions,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    },
    login: jest.fn(),
  };
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [storageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${storageAclUrl}>; rel="acl"`,
        }),
      [storageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export function mockAuthenticatedSessionWithNoAccessToAnotherUsersPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [anotherUsersStorageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${anotherUsersStorageAclUrl}>; rel="acl"`,
        }),
      [anotherUsersStorageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export const mockAuthenticatedSession =
  mockAuthenticatedSessionWithNoAccessToPod;
