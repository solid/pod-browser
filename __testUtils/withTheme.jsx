/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
