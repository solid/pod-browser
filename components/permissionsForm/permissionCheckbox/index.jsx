/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import T from "prop-types";
import {
  Checkbox,
  FormControlLabel,
  ListItem,
  makeStyles,
  createStyles,
} from "@mui/core";
import styles from "../styles";

const TESTCAFE_ID_PERMISSION_LABEL = "permission-label-";
export const TESTCASE_ID_PERMISSION_CHECKBOX = "permission-checkbox-";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

function PermissionCheckbox({ value, label, disabled, onChange }) {
  const classes = useStyles();
  const name = label.toLowerCase();

  return (
    // prettier-ignore
    <ListItem className={classes.listItem}>
      <FormControlLabel
        classes={{ label: classes.label }}
        data-testid={TESTCAFE_ID_PERMISSION_LABEL + name}
        label={label}
        control={(
          <Checkbox
            classes={{ root: classes.checkbox }}
            checked={value}
            name={name}
            onChange={onChange}
            disabled={disabled}
            inputProps={{"data-testid": TESTCASE_ID_PERMISSION_CHECKBOX + name}}
          />
        )}
      />
    </ListItem>
  );
}

PermissionCheckbox.propTypes = {
  disabled: T.bool,
  value: T.bool.isRequired,
  label: T.string.isRequired,
  onChange: T.func,
};

PermissionCheckbox.defaultProps = {
  disabled: false,
  onChange: () => {},
};

export default PermissionCheckbox;
