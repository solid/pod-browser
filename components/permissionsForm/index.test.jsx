/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable camelcase */
import React from "react";
import * as routerFns from "next/router";
import PermissionsForm, {
  arrowIcon,
  permissionHandler,
  toggleOpen,
} from "./index";
import { ACL, createAccessMap } from "../../src/solidClientHelpers/permissions";
import { renderWithTheme } from "../../__testUtils/withTheme";
import useFullProfile from "../../src/hooks/useFullProfile";
import { mockProfileAlice } from "../../__testUtils/mockPersonResource";
import { TESTCASE_ID_PERMISSION_CHECKBOX } from "./permissionCheckbox";
import { joinPath } from "../../src/stringHelpers";
import mockSession from "../../__testUtils/mockSession";
import mockSessionContextProvider from "../../__testUtils/mockSessionContextProvider";

jest.mock("../../src/hooks/useFullProfile");
const mockedUseFullProfile = useFullProfile;

describe("PermissionsForm", () => {
  const authProfile = mockProfileAlice();
  const authOwnedIri = joinPath(authProfile.pods[0], "test");
  const visitorIri = "http://some-random-pod.com/";
  const viewCheckbox = `${TESTCASE_ID_PERMISSION_CHECKBOX}view`;
  const editCheckbox = `${TESTCASE_ID_PERMISSION_CHECKBOX}edit`;
  const appendCheckbox = `${TESTCASE_ID_PERMISSION_CHECKBOX}append`;
  const controlCheckbox = `${TESTCASE_ID_PERMISSION_CHECKBOX}control`;
  const acl = createAccessMap(true, true, true, true);

  let mockedRouterHook;
  let SessionProvider;

  beforeEach(() => {
    mockedRouterHook = jest
      .spyOn(routerFns, "useRouter")
      .mockReturnValue({ query: { iri: visitorIri } });
    mockedUseFullProfile.mockReturnValue(authProfile);
  });
  beforeEach(() => {
    const session = mockSession({ webId: authProfile.webId });
    SessionProvider = mockSessionContextProvider(session);
  });

  it("Renders a permissions form", () => {
    const { asFragment, getByTestId } = renderWithTheme(
      <PermissionsForm acl={acl} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(viewCheckbox).disabled).toBe(false);
    expect(getByTestId(editCheckbox).disabled).toBe(false);
    expect(getByTestId(appendCheckbox).disabled).toBe(false);
    expect(getByTestId(controlCheckbox).disabled).toBe(false);
  });

  it("returns null if control is false", () => {
    const aclNoControl = createAccessMap(true, true, true, false);

    const { asFragment } = renderWithTheme(
      <PermissionsForm acl={aclNoControl} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("toggleOpen", () => {
    let setOpen;

    beforeEach(() => {
      setOpen = jest.fn();
    });

    it("calls setOpen with false is open is true", () => {
      toggleOpen(true, setOpen)();

      expect(setOpen).toHaveBeenCalledWith(false);
    });

    it("calls setOpen with true is open is false", () => {
      toggleOpen(false, setOpen)();

      expect(setOpen).toHaveBeenCalledWith(true);
    });
  });

  describe("arrowIcon", () => {
    it("returns the up arrow when open is true", () => {
      const { asFragment } = renderWithTheme(arrowIcon(true));
      expect(asFragment()).toMatchSnapshot();
    });

    it("returns the down arrow when open is false", () => {
      const { asFragment } = renderWithTheme(arrowIcon(false));
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("can be disabled explicitly", () => {
    const { getByTestId } = renderWithTheme(
      <PermissionsForm acl={acl} disabled />
    );
    expect(getByTestId(viewCheckbox).disabled).toBe(true);
    expect(getByTestId(editCheckbox).disabled).toBe(true);
    expect(getByTestId(appendCheckbox).disabled).toBe(true);
    expect(getByTestId(controlCheckbox).disabled).toBe(true);
  });

  it("is disabled connected to authenticated user and the resource is part of their Pod", () => {
    mockedRouterHook.mockReturnValue({ query: { iri: authOwnedIri } });
    const { getByTestId } = renderWithTheme(
      <SessionProvider>
        <PermissionsForm acl={acl} webId={authProfile.webId} />
      </SessionProvider>
    );
    expect(getByTestId(viewCheckbox).disabled).toBe(true);
    expect(getByTestId(editCheckbox).disabled).toBe(true);
    expect(getByTestId(appendCheckbox).disabled).toBe(true);
    expect(getByTestId(controlCheckbox).disabled).toBe(true);
  });
});

describe("permissionHandler", () => {
  it("updates access map", () => {
    const access = createAccessMap(true, true);
    const setAccess = jest.fn();
    const onChange = jest.fn();

    const setPermissionHandler = permissionHandler(access, setAccess, onChange);
    setPermissionHandler(ACL.READ.key)();

    expect(setAccess).toHaveBeenCalled();
    expect(onChange).toHaveBeenCalled();
  });
});
