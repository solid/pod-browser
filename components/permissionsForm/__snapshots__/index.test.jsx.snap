// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsForm Renders a permissions form 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-container"
  >
    <button
      class="PodBrowser-root PodBrowser-root PodBrowser-text PodBrowser-summary"
      data-testid="permissions-dropdown-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <span>
          Control
        </span>
        <span
          class="PodBrowser-endIcon PodBrowser-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="PodBrowser-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
        </span>
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <section
      class="PodBrowser-selectionClosed"
    >
      <ul
        class="PodBrowser-root PodBrowser-padding"
      >
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-view"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-view"
                  name="view"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              View
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-edit"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-edit"
                  name="edit"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Edit
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-append"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-append"
                  name="append"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Append
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-control"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-control"
                  name="control"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Control
            </span>
          </label>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;

exports[`PermissionsForm arrowIcon returns the down arrow when open is false 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="PodBrowser-root"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
    />
  </svg>
</DocumentFragment>
`;

exports[`PermissionsForm arrowIcon returns the up arrow when open is true 1`] = `
<DocumentFragment>
  <svg
    aria-hidden="true"
    class="PodBrowser-root"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
    />
  </svg>
</DocumentFragment>
`;

exports[`PermissionsForm returns null if control is false 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-container"
  >
    <button
      class="PodBrowser-root PodBrowser-root PodBrowser-text PodBrowser-summary"
      data-testid="permissions-dropdown-button"
      tabindex="0"
      type="button"
    >
      <span
        class="PodBrowser-label"
      >
        <span>
          Edit
        </span>
        <span
          class="PodBrowser-endIcon PodBrowser-iconSizeMedium"
        >
          <svg
            aria-hidden="true"
            class="PodBrowser-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            />
          </svg>
        </span>
      </span>
      <span
        class="PodBrowser-root"
      />
    </button>
    <section
      class="PodBrowser-selectionClosed"
    >
      <ul
        class="PodBrowser-root PodBrowser-padding"
      >
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-view"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-view"
                  name="view"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              View
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-edit"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-edit"
                  name="edit"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Edit
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-append"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-checked PodBrowser-checked PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  checked=""
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-append"
                  name="append"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Append
            </span>
          </label>
        </li>
        <li
          class="PodBrowser-root PodBrowser-listItem PodBrowser-gutters"
        >
          <label
            class="PodBrowser-root"
            data-testid="permission-label-control"
          >
            <span
              aria-disabled="false"
              class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-checkbox PodBrowser-colorSecondary PodBrowser-colorSecondary"
            >
              <span
                class="PodBrowser-label"
              >
                <input
                  class="PodBrowser-input"
                  data-indeterminate="false"
                  data-testid="permission-checkbox-control"
                  name="control"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="PodBrowser-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="PodBrowser-root"
              />
            </span>
            <span
              class="PodBrowser-root PodBrowser-label PodBrowser-label PodBrowser-body1"
            >
              Control
            </span>
          </label>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
